lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -1004,6 +1022,116 @@ struct BroadcastOpConversion
   }
 };
 
+struct ReduceOpConversion
+    : public ConvertTritonGPUOpToLLVMPattern<triton::ReduceOp> {
+
+  using ConvertTritonGPUOpToLLVMPattern<
+      triton::ReduceOp>::ConvertTritonGPUOpToLLVMPattern;
+
+  LogicalResult
+  matchAndRewrite(triton::ReduceOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    Location loc = op->getLoc();
+    unsigned axis = adaptor.axis();
+
+    auto do_acc = [&](Value &acc, Value cur, bool is_first) -> void {
+      if (is_first) {
+        acc = cur;
+      } else {
+        acc = rewriter.create<LLVM::FAddOp>(loc, acc.getType(), acc, cur);
+      }
+    };
+
+    auto srcTy = op.operand().getType().template dyn_cast<RankedTensorType>();
+    auto srcLayout = srcTy.getEncoding().dyn_cast<BlockedEncodingAttr>();
+    auto srcShape = srcTy.getShape();
+
+    auto llvmElemTy = getTypeConverter()->convertType(srcTy.getElementType());
+    auto elemPtrTy = LLVM::LLVMPointerType::get(llvmElemTy, 3);
+    Value smemBase = getSharedMemoryBase(loc, rewriter, op.getOperation());
+    smemBase = bitcast(elemPtrTy, smemBase);
+
+    std::vector<unsigned> shmShape;
+    for (auto d : srcShape)
+      shmShape.push_back(d);
+    shmShape[axis] =
+        srcLayout.getThreadsPerWarp()[axis] * srcLayout.getWarpsPerCTA()[axis];
+
+    unsigned srcElems = getElemsPerThread(srcLayout, srcShape);
+    auto srcIndices = emitIndices(loc, rewriter, srcLayout, srcShape);
+    auto srcValues =
+        getElementsFromStruct(loc, adaptor.operand(), srcElems, rewriter);
+
+    SmallVector<SmallVector<unsigned>> offset =
+        emitOffsetForBlockedLayout(srcLayout, srcShape);
+
+    std::map<SmallVector<unsigned>, Value> accs;
+    std::map<SmallVector<unsigned>, SmallVector<Value>> indices;
+
+    for (int i = 0; i < srcElems; ++i) {
+      SmallVector<unsigned> key = offset[i];
+      key[axis] = 0;
+      bool is_first = accs.find(key) == accs.end();
+      do_acc(accs[key], srcValues[i], is_first);
+      if (is_first) {
+        indices[key] = srcIndices[i];
+      }
+    }
+
+    Value zero = i32_val(0);
+
+    for (auto it : accs) {
+      const SmallVector<unsigned> &key = it.first;
+      Value acc = it.second;
+      SmallVector<Value> writeIdx = indices[key];
+
+      writeIdx[axis] =
+          udiv(writeIdx[axis], i32_val(srcLayout.getSizePerThread()[axis]));
```

Comment:
Just a suggestion, it's better to put the `i32_val` outside the forloop, or the pass will put many redundant `llvm.constant` for the same constant number to the IR, not affect the final performance, but a little hard to read the immediate IR after it.

Reply:
Thanks for this suggestion. Code just updated.